### PR TITLE
Include prompt templates in genai_config.json in model builder

### PR DESCRIPTION
**Pull request by vortex-captain**
This PR aims to generate optionally prompt templates for different roles in model builder. These templates are required for some downstream frameworks to serve chats correctly.

Example: 
(This option is enabled by `--extra_options prompt_templates=1` in model builder CLI tool)
```powershell
python -m onnxruntime_genai.models.builder -m microsoft/Phi-3-mini-4k-instruct -p int4 -e cpu -o <output_path> -c <cache_path> --extra_options prompt_templates=1
```

will save `genai_config.json` with these templates:
```json
{
    "model": {
        ...
        "prompt_templates": {
            "assistant": "<|assistant|>\n{Content}<|end|>\n",
            "prompt": "<|user|>\n{Content}<|end|>\n<|assistant|>\n",
            "system": "<|system|>\n{Content}<|end|>\n",
            "user": "<|user|>\n{Content}<|end|>\n"
        },
        ...
    },
    ...
}
```

### CODE REVIEW

**Comment by mindest** [comment]
@vortex-captain Pls help review the request above

**Comment by vortex-captain** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: src/python/py/models/builder.py [hunk 1]
```diff
@@ -364,15 +364,44 @@ def make_genai_config(self, model_name_or_path, extra_kwargs, out_dir):
...
+            prompt_template = prompt_template[:prompt_template.rfind('{Content}')]
+            templates = {
+                "system": system_template,
+                "user": user_template,
+                "assistant": assistant_template,
+                "prompt": prompt_template
+            }
+            return templates 
+        except Exception as e:
+            print(f"Failed to save inference_model.json. Error: {e}")
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builder.py:402]:
```suggestion
            print(f"Failed to get prompt templates. Error: {e}")
```

File: src/python/py/models/builder.py [hunk 2]
```diff
@@ -364,15 +364,44 @@ def make_genai_config(self, model_name_or_path, extra_kwargs, out_dir):
...
         tokenizer.save_pretrained(out_dir)
 
+    def _get_prompt_templates(self, hf_name, extra_kwargs):
+        try:
+            # disable end of sentence padding with eos_token=None
+            tokenizer = AutoTokenizer.from_pretrained(hf_name, token=self.hf_token, trust_remote_code=True, eos_token=None, **extra_kwargs)
+            system_template = tokenizer.apply_chat_template([{'role': 'system', 'content': '{Content}'}], tokenize=False)
+            system_user_template = tokenizer.apply_chat_template([{'role': 'system', 'content': '{Content}'}, {'role': 'user', 'content': '{Content}'}], tokenize=False)
+            system_user_assistant_template = tokenizer.apply_chat_template([{'role': 'system', 'content': '{Content}'}, {'role': 'user', 'content': '{Content}'}, {'role': 'assistant', 'content': '{Content}'}], tokenize=False)
+            assert system_user_template.startswith(system_template), "chat templates may contain padding tokens, leading to incorrect inference_model.json"
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builder.py:388]:
```suggestion
            assert system_user_template.startswith(system_template), "Chat templates may contain padding tokens, leading to incorrect prompt templates"
```

File: src/python/py/models/builder.py [hunk 3]
```diff
@@ -364,15 +364,44 @@ def make_genai_config(self, model_name_or_path, extra_kwargs, out_dir):
...
+            assert system_user_assistant_template.startswith(system_user_template), "chat templates may contain padding tokens, leading to incorrect inference_model.json"
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builder.py:389]:
```suggestion
            assert system_user_assistant_template.startswith(system_user_template), "Chat templates may contain padding tokens, leading to incorrect prompt templates"
```

**Review by kunal-vaishnavi — APPROVED** [review]

**Pull request merged**